erts/emulator/nifs/common/prim_file_nif.c

```diff
@@ -1399,3 +1402,32 @@ static ERL_NIF_TERM altname_nif(ErlNifEnv *env, int argc, const ERL_NIF_TERM arg
 
     return enif_make_tuple2(env, am_ok, result);
 }
+
+static ERL_NIF_TERM copy_file_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]) {
+    posix_errno_t posix_errno;
+    off64_t ret = 0;
+    efile_path_t file_path_in, file_path_out;
+
+    ASSERT(argc == 2);
+
+    if ((posix_errno = efile_marshal_path(env, argv[0], &file_path_in))) {
+        return posix_error_to_tuple(env, posix_errno);
+    }
+
+    if ((posix_errno = efile_marshal_path(env, argv[1], &file_path_out))) {
+        return posix_error_to_tuple(env, posix_errno);
+    }
+
+#ifdef HAVE_COPY_FILE_RANGE
+    /* Copy file contents using unix SYS_copy_file_range syscall */
+    if ((posix_errno = efile_copy_file_range(&file_path_in, &file_path_out, &ret))) {
+        return posix_error_to_tuple(env, posix_errno);
+    }
+ #else 
+    /* As a fallback for when copy_file_range(2) is unavailable wecould do a simple read(2)+write(2) loop. */
+    if ((posix_errno = efile_simple_copy_file(&file_path_in, &file_path_out, &ret))) {
+        return posix_error_to_tuple(env, posix_errno);
+    }
+ #endif   
+    return enif_make_tuple2(env, enif_make_atom(env, "ok"), enif_make_uint(env, (unsigned int)ret));
```

Comment:
Returning as 'am_ok' now. However, both `file:copy/2` and `file:copy/3` return {ok, BytesCopied}. So, why are we only returning 'ok'? Considering the user uses `copy_range/3`, we can return the same length that they passed into the function. But what should we do in the case of `copy_file/2` when only Source and Destination passed?

Reply:
Ah, sorry, I forgot that `file:copy/2` is already documented as `{ok, BytesCopied}` -- it wasn't introduced by this PR.

> But what should we do in the case of copy_file/2 when only Source and Destination passed?

Hm, I think `file:copy/2` should either do a range copy (when available) or use the fallback. The current interface of   `file:copy/2` means that we cannot use `CopyFileW` or equivalent, so a separate function may be warranted.

I plan on refactoring the current `file` interface into a new module that has all our lessons learned baked into a cleaner interface (no idea when though), so we can either live with this until then or forget about `CopyFileW` for now, I'll leave that up to you. :-)

Reply:
What the reason not to use `CopyFileW` in case of `file:copy/2`? In case of `Filename` as arguments we can. In case of  `pid()` or `fd()` we cant use it, but we can fallback to already implemented `copy` in erlang? Also for unix `copy_file/2` and `copy_range/3`  handling only `Filenames` now, need to handle `fd()` and `pid()` as well to match `file/copy/{2,3}` . Don't care much about `CopyFileW`, doing this just to increase performance of unix systems file copy, but if find out a way to handle `pid()`, `fd()`, and `Filename` as params for `copy_file/2`, then for sure will figure something out for Windows

Reply:
> What's the reason for not using CopyFileW in the case of file:copy/2?

Not getting an accurate count of bytes copied. We can only check before and after, and it might have changed in the meantime.

> However, when using pid() or fd(), we can't utilize it. Nevertheless, we can fallback to the already implemented copy in Erlang, right? Moreover, for Unix systems, copy_file/2 and copy_range/3 currently handle only Filenames. There's a need to extend support to handle fd() and pid() as well, aligning with the functionality of file/copy/{2,3}.

`fd()` and `pid()` involve communicating with other processes that might not be wrapping a file in the first place. It's better to just leave all that to the Erlang fallback.